types/async-eventemitter/index.d.ts

```diff
@@ -0,0 +1,107 @@
+   */
+  emit<E extends keyof T>(
+    event: E & string,
+    ...args: Parameters<T[E]>
+  ): boolean;
+  /**
+   * Adds a listener to the beginning of the listeners array for the specified event.
+   * @param event EventMap key (event name)
+   * @param listener EventMap value (event function)
+   * @see https://www.npmjs.com/package/async-eventemitter#important-differences-between-asynceventemitter-the-native-eventemitter
+   */
+  first<E extends keyof T>(event: E & string, listener: T[E]): this;
+  /**
+   * Adds a listener at the specified index in the listeners array for the specified event.
+   * @param event EventMap key (event name)
+   * @param index Index to insert at
+   * @param listener EventMap value (event function)
+   * @see https://www.npmjs.com/package/async-eventemitter#important-differences-between-asynceventemitter-the-native-eventemitter
+   */
+  at<E extends keyof T>(event: E & string, index: number, listener: T[E]): this;
+  /**
+   * Adds a listener before the target listener in the listeners array for the specified event.
+   * @param event EventMap key (event name)
+   * @param target Listener to insert before
+   * @param listener EventMap value (event function)
+   * @see https://www.npmjs.com/package/async-eventemitter#important-differences-between-asynceventemitter-the-native-eventemitter
+   */
+  before<E extends keyof T>(
+    event: E & string,
+    target: T[E],
+    listener: T[E],
+  ): this;
+  /**
+   * Adds a listener after the target listener in the listeners array for the specified event.
+   * @param event EventMap key (event name)
+   * @param target Listener to insert before
+   * @param listener EventMap value (event function)
+   * @see https://www.npmjs.com/package/async-eventemitter#important-differences-between-asynceventemitter-the-native-eventemitter
+   */
+  after<E extends keyof T>(
+    event: E & string,
+    target: T[E],
+    listener: T[E],
+  ): this;
+
+  // https://github.com/andywer/typed-emitter/blob/master/index.d.ts
+  addListener<E extends keyof T>(event: E & string, listener: T[E]): this;
+  on<E extends keyof T>(event: E & string, listener: T[E]): this;
+  once<E extends keyof T>(event: E & string, listener: T[E]): this;
+  prependListener<E extends keyof T>(event: E & string, listener: T[E]): this;
+  prependOnceListener<E extends keyof T>(
+    event: E & string,
+    listener: T[E],
+  ): this;
+
+  removeAllListeners(event?: keyof T & string): this;
+  removeListener<E extends keyof T>(event: E & string, listener: T[E]): this;
+
+  eventNames(): Array<keyof T & string>;
+  listeners<E extends keyof T>(event: E & string): Array<T[E]>;
+  listenerCount(event: keyof T & string): number;
+
+  getMaxListeners(): number;
+  setMaxListeners(maxListeners: number): this;
+}
+
+declare namespace AsyncEventEmitter {
+    type AsyncListener<T, R> = (data?: T, callback?: (result?: R) => void) => Promise<R> | void;
```

Comment:
Also, the type `Promise<R> | void` doesn't really make sense. `void` only makes sense as the only return type of a function.

This should be the following:
```ts
type AsyncListener<T, R> = ((data: T, callback: (result?: R) => void) => Promise<R>) | ((data: T, callback: (result?: R) => void) => void);
```
That is, either it's listener returns nothing, or it returns something, which is a promise of `R`.